File: app/src/main/java/com/rollapp/shared/ui/auth/AuthScreen.kt
package com.rollapp.shared.ui.auth

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Visibility
import androidx.compose.material.icons.rounded.VisibilityOff
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.rollapp.shared.ui.components.InlineError

@Composable
fun AuthScreen(
    onSignedIn: () -> Unit,
    viewModel: AuthViewModel = hiltViewModel()
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val context = LocalContext.current
    val snackbarHostState = remember { SnackbarHostState() }
    var passwordVisible by remember { mutableStateOf(false) }

    LaunchedEffect(state.signedIn) {
        if (state.signedIn) onSignedIn()
    }

    LaunchedEffect(state.passwordResetSent) {
        if (state.passwordResetSent) {
            snackbarHostState.showSnackbar("Check your inbox for a reset link")
        }
    }

    Scaffold(
        snackbarHost = { SnackbarHost(snackbarHostState) }
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .verticalScroll(rememberScrollState())
                .imePadding()
                .padding(horizontal = 28.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Spacer(Modifier.height(56.dp))

            Text(
                text = "Roll",
                style = MaterialTheme.typography.displaySmall,
                color = MaterialTheme.colorScheme.primary
            )
            Spacer(Modifier.height(8.dp))
            Text(
                text = "One shared camera roll for the people you were with.",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center
            )

            Spacer(Modifier.height(40.dp))

            AnimatedVisibility(visible = state.mode == AuthMode.SIGN_UP) {
                Column {
                    OutlinedTextField(
                        value = state.name,
                        onValueChange = viewModel::onNameChange,
                        label = { Text("Your name") },
                        singleLine = true,
                        shape = MaterialTheme.shapes.medium,
                        keyboardOptions = KeyboardOptions(imeAction = ImeAction.Next),
                        modifier = Modifier.fillMaxWidth()
                    )
                    Spacer(Modifier.height(12.dp))
                }
            }

            OutlinedTextField(
                value = state.email,
                onValueChange = viewModel::onEmailChange,
                label = { Text("Email") },
                singleLine = true,
                shape = MaterialTheme.shapes.medium,
                keyboardOptions = KeyboardOptions(
                    keyboardType = KeyboardType.Email,
                    imeAction = ImeAction.Next
                ),
                modifier = Modifier.fillMaxWidth()
            )

            Spacer(Modifier.height(12.dp))

            OutlinedTextField(
                value = state.password,
                onValueChange = viewModel::onPasswordChange,
                label = { Text("Password") },
                singleLine = true,
                shape = MaterialTheme.shapes.medium,
                visualTransformation = if (passwordVisible) VisualTransformation.None
                else PasswordVisualTransformation(),
                keyboardOptions = KeyboardOptions(
                    keyboardType = KeyboardType.Password,
                    imeAction = ImeAction.Done
                ),
                trailingIcon = {
                    IconButton(onClick = { passwordVisible = !passwordVisible }) {
                        Icon(
                            imageVector = if (passwordVisible) Icons.Rounded.VisibilityOff
                            else Icons.Rounded.Visibility,
                            contentDescription = if (passwordVisible) "Hide password" else "Show password"
                        )
                    }
                },
                modifier = Modifier.fillMaxWidth()
            )

            if (state.mode == AuthMode.SIGN_IN) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.End
                ) {
                    TextButton(
                        onClick = viewModel::sendPasswordReset,
                        enabled = state.email.contains("@")
                    ) { Text("Forgot password?") }
                }
            } else {
                Spacer(Modifier.height(12.dp))
            }

            state.error?.let { error ->
                InlineError(
                    message = error.message ?: "Something went wrong",
                    modifier = Modifier.padding(bottom = 12.dp)
                )
            }

            Button(
                onClick = viewModel::submit,
                enabled = state.canSubmit,
                shape = MaterialTheme.shapes.medium,
                contentPadding = ButtonDefaults.ContentPadding,
                modifier = Modifier
                    .fillMaxWidth()
                    .height(52.dp)
            ) {
                if (state.isSubmitting) {
                    CircularProgressIndicator(
                        strokeWidth = 2.dp,
                        modifier = Modifier.size(20.dp),
                        color = MaterialTheme.colorScheme.onPrimary
                    )
                } else {
                    Text(
                        text = if (state.mode == AuthMode.SIGN_IN) "Sign in" else "Create account",
                        style = MaterialTheme.typography.labelLarge
                    )
                }
            }

            Spacer(Modifier.height(20.dp))

            Row(verticalAlignment = Alignment.CenterVertically) {
                HorizontalDivider(modifier = Modifier.weight(1f))
                Text(
                    text = "or",
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.padding(horizontal = 12.dp)
                )
                HorizontalDivider(modifier = Modifier.weight(1f))
            }

            Spacer(Modifier.height(20.dp))

            OutlinedButton(
                onClick = { viewModel.signInWithGoogle(context) },
                enabled = !state.isSubmitting,
                shape = MaterialTheme.shapes.medium,
                modifier = Modifier
                    .fillMaxWidth()
                    .height(52.dp)
            ) {
                Text("Continue with Google", style = MaterialTheme.typography.labelLarge)
            }

            Spacer(Modifier.height(10.dp))

            TextButton(
                onClick = viewModel::continueAsGuest,
                enabled = !state.isSubmitting,
                modifier = Modifier.fillMaxWidth()
            ) {
                Text("Just looking — continue as guest")
            }

            Spacer(Modifier.height(16.dp))

            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.Center
            ) {
                Text(
                    text = if (state.mode == AuthMode.SIGN_IN) "New here?" else "Already have an account?",
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Spacer(Modifier.width(4.dp))
                TextButton(
                    onClick = {
                        viewModel.setMode(
                            if (state.mode == AuthMode.SIGN_IN) AuthMode.SIGN_UP else AuthMode.SIGN_IN
                        )
                    }
                ) {
                    Text(if (state.mode == AuthMode.SIGN_IN) "Create an account" else "Sign in")
                }
            }

            Spacer(Modifier.height(32.dp))
        }
    }
}

File: ios/WeWere/UI/Screens/AuthScreen.swift
import SwiftUI
import Combine
import FirebaseCore
import GoogleSignIn

enum AuthMode { case signIn, signUp }

struct AuthUiState: Equatable {
    var mode: AuthMode = .signIn
    var name: String = ""
    var email: String = ""
    var password: String = ""
    var isSubmitting: Bool = false
    var error: AppError? = nil
    var passwordResetSent: Bool = false
    var signedIn: Bool = false

    var canSubmit: Bool {
        !isSubmitting && email.contains("@") && password.count >= 6 &&
            (mode == .signIn || !name.trimmingCharacters(in: .whitespaces).isEmpty)
    }
}

@MainActor
final class AuthViewModel: ObservableObject {
    @Published var state = AuthUiState()

    private let authRepository: AuthRepository

    init(container: AppContainer = .shared) {
        authRepository = container.authRepository
    }

    func setMode(_ mode: AuthMode) { state.mode = mode; state.error = nil }
    func onNameChange(_ value: String) { state.name = value; state.error = nil }
    func onEmailChange(_ value: String) { state.email = value; state.error = nil }
    func onPasswordChange(_ value: String) { state.password = value; state.error = nil }
    func dismissError() { state.error = nil }

    func submit() {
        let current = state
        guard current.canSubmit else { return }
        Task {
            state.isSubmitting = true; state.error = nil
            let outcome: Outcome<User>
            switch current.mode {
            case .signIn: outcome = await authRepository.signInWithEmail(email: current.email, password: current.password)
            case .signUp: outcome = await authRepository.signUpWithEmail(name: current.name, email: current.email, password: current.password)
            }
            apply(outcome.unit)
        }
    }

    func signInWithGoogle() {
        Task {
            state.isSubmitting = true; state.error = nil
            switch await GoogleSignInHelper.getTokens() {
            case .failure(let error):
                state.isSubmitting = false; state.error = error
            case .success(let tokens):
                apply(await authRepository.signInWithGoogle(idToken: tokens.idToken, accessToken: tokens.accessToken).unit)
            }
        }
    }

    func continueAsGuest() {
        Task {
            state.isSubmitting = true; state.error = nil
            apply(await authRepository.signInAnonymously().unit)
        }
    }

    func sendPasswordReset() {
        let email = state.email
        Task {
            state.isSubmitting = true; state.error = nil
            switch await authRepository.sendPasswordReset(email: email) {
            case .success: state.isSubmitting = false; state.passwordResetSent = true
            case .failure(let error): state.isSubmitting = false; state.error = error
            }
        }
    }

    private func apply(_ outcome: Outcome<Void>) {
        switch outcome {
        case .success: state.isSubmitting = false; state.signedIn = true
        case .failure(let error): state.isSubmitting = false; state.error = error
        }
    }
}

/// Google sign-in through the Google Sign-In SDK, presented over the current screen.
@MainActor
enum GoogleSignInHelper {
    static func getTokens() async -> Outcome<(idToken: String, accessToken: String)> {
        guard let clientID = FirebaseApp.app()?.options.clientID, !clientID.isEmpty else {
            return .failure(.validation("Google Sign-In isn't configured — add GoogleService-Info.plist"))
        }
        GIDSignIn.sharedInstance.configuration = GIDConfiguration(clientID: clientID)
        guard let presenter = UIApplication.topViewController() else {
            return .failure(.unknown("Nothing to present sign-in from"))
        }
        do {
            let result = try await GIDSignIn.sharedInstance.signIn(withPresenting: presenter)
            guard let idToken = result.user.idToken?.tokenString else {
                return .failure(.unknown("Google sign-in returned no token"))
            }
            return .success((idToken, result.user.accessToken.tokenString))
        } catch {
            let ns = error as NSError
            if ns.domain == kGIDSignInErrorDomain && ns.code == GIDSignInError.Code.canceled.rawValue {
                return .failure(.validation("Sign-in cancelled"))
            }
            return .failure(.unknown(ns.localizedDescription))
        }
    }
}

struct AuthScreen: View {
    let onSignedIn: () -> Void

    @StateObject private var viewModel = AuthViewModel()
    @EnvironmentObject private var overlays: Overlays
    @State private var passwordVisible = false
    // Google is the front door; the email form only unfolds when asked for.
    @State private var emailFormOpen = false

    var body: some View {
        let state = viewModel.state
        GeometryReader { geo in
        ScrollView {
            VStack(spacing: 0) {
                if !emailFormOpen {
                    PrintStack()
                        .frame(height: 360)
                        .frame(maxWidth: .infinity)
                        .transition(.opacity.combined(with: .move(edge: .top)))
                } else {
                    Spacer().frame(height: 72)
                }

                RiseIn(delay: 0.5) {
                    Text("WeWere").rollStyle(RollType.displayLarge, color: Gold)
                }
                Spacer().frame(height: 10)
                RiseIn(delay: 0.6) { HairlineRule().frame(width: 120) }
                Spacer().frame(height: 10)
                RiseIn(delay: 0.65) {
                    Text("One shared camera roll for the people you were with.")
                        .rollStyle(RollType.bodyMedium, color: IvoryMuted)
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, 56)
                }

                Spacer(minLength: 32)

                VStack(spacing: 12) {
                    if let error = state.error {
                        InlineError(message: error.message ?? "Something went wrong")
                    }

                    if emailFormOpen {
                        EmailForm(passwordVisible: $passwordVisible, viewModel: viewModel)
                            .transition(.opacity.combined(with: .move(edge: .bottom)))
                        QuietButton(text: "Back to Google sign-in", enabled: !state.isSubmitting) {
                            withAnimation(.settle(duration: 0.4)) { emailFormOpen = false }
                        }
                    } else {
                        RiseIn(delay: 0.78) {
                            GoldButton(text: "Continue with Google", enabled: !state.isSubmitting, loading: state.isSubmitting) {
                                viewModel.signInWithGoogle()
                            }
                        }
                        RiseIn(delay: 0.86) {
                            HairlineButton(text: "Use email instead", enabled: !state.isSubmitting) {
                                withAnimation(.settle(duration: 0.4)) { emailFormOpen = true }
                            }
                        }
                        RiseIn(delay: 0.94) {
                            QuietButton(text: "Just looking — continue as guest", enabled: !state.isSubmitting) {
                                viewModel.continueAsGuest()
                            }
                        }
                    }
                }
                .padding(.horizontal, 24)

                Spacer().frame(height: 16)
            }
            .frame(minHeight: geo.size.height)
        }
        .scrollBounceBehavior(.basedOnSize)
        }
        .background(
            ZStack {
                Ink
                RadialGradient(colors: [Color(hex: 0x1A1712), Ink], center: .topLeading, startRadius: 0, endRadius: 500)
            }
            .ignoresSafeArea()
        )
        .onChange(of: state.signedIn) { _, signedIn in if signedIn { onSignedIn() } }
        .onChange(of: state.passwordResetSent) { _, sent in
            if sent { overlays.snack("Check your inbox for a reset link") }
        }
    }
}

private struct EmailForm: View {
    @Binding var passwordVisible: Bool
    @ObservedObject var viewModel: AuthViewModel

    var body: some View {
        let state = viewModel.state
        VStack(spacing: 12) {
            if state.mode == .signUp {
                RollTextField(label: "Your name",
                              text: Binding(get: { viewModel.state.name }, set: viewModel.onNameChange),
                              capitalization: .words, submitLabel: .next)
            }

            RollTextField(label: "Email",
                          text: Binding(get: { viewModel.state.email }, set: viewModel.onEmailChange),
                          keyboard: .emailAddress, capitalization: .never, submitLabel: .next)

            RollTextField(label: "Password",
                          text: Binding(get: { viewModel.state.password }, set: viewModel.onPasswordChange),
                          isSecure: !passwordVisible, capitalization: .never, submitLabel: .done,
                          onSubmit: { viewModel.submit() }) {
                IconButton(icon: passwordVisible ? RollIcon.visibilityOff : RollIcon.visibility, tint: Muted, size: 22,
                           label: passwordVisible ? "Hide password" : "Show password") {
                    passwordVisible.toggle()
                }
            }

            if state.mode == .signIn {
                HStack {
                    Spacer()
                    TextButton(text: "Forgot password?", color: Gold, style: RollType.bodySmall,
                               enabled: state.email.contains("@")) {
                        viewModel.sendPasswordReset()
                    }
                }
            }

            GoldButton(text: state.mode == .signIn ? "Sign in" : "Create account",
                       enabled: state.canSubmit, loading: state.isSubmitting) {
                viewModel.submit()
            }

            HStack(spacing: 0) {
                Text(state.mode == .signIn ? "New here?" : "Already have an account?")
                    .rollStyle(RollType.bodySmall, color: Muted)
                TextButton(text: state.mode == .signIn ? "Create an account" : "Sign in",
                           color: Gold, style: RollType.bodySmall) {
                    withAnimation(.easeOut(duration: 0.2)) {
                        viewModel.setMode(state.mode == .signIn ? .signUp : .signIn)
                    }
                }
            }
            .frame(maxWidth: .infinity)
        }
    }
}
